File: crates/normalize-languages/tests/fixtures/java/variants.java
// Completeness fixture: one construct per grammar-legal variant of each field
// the java.{tags,calls,imports,complexity,types}.scm queries constrain,
// cross-referenced against arborium-java 2.17.0's node-types.json. Every
// construct here is *expected to be captured*; see query_fixtures.rs
// `java_*_completeness_*` tests for the matrix.
//
// This file also carries a NEGATIVE section with deliberate near-miss
// constructs that must NOT be captured by the query under test, to guard
// against over-broad patterns.

import Bare; // import_declaration argument: bare `identifier` (no package) —
             // rare/non-idiomatic (default-package imports aren't legal Java)
             // but grammar-legal; see java.imports.scm.
import java.util.ArrayList; // import_declaration argument: scoped_identifier
import java.util.List; // plain multi-segment import
import java.util.*; // import_declaration: scoped_identifier + asterisk (wildcard)
import static java.lang.Math.PI; // import static pkg.Class.member;
import static java.lang.Math.*; // import static pkg.Class.*;

// --- object_creation_expression.type variants -------------------------------

class NewVariants {
    void plainNew() {
        new Object(); // type: type_identifier
    }

    void genericNew() {
        new ArrayList<String>(); // type: generic_type -> type_identifier
    }

    void genericDiamondNew() {
        new ArrayList<>(); // type: generic_type -> type_identifier (diamond)
    }

    void scopedNew() {
        new java.util.Date(); // type: scoped_type_identifier
    }

    void genericScopedNew() {
        new java.util.HashMap<String, Integer>(); // type: generic_type -> scoped_type_identifier
    }
}

// --- superclass variants ----------------------------------------------------

class PlainBase {}

class GenericBase<T> {}

class ExtendsPlain extends PlainBase {} // superclass: type_identifier

class ExtendsGeneric extends GenericBase<String> {} // superclass: generic_type -> type_identifier

class ExtendsScoped extends java.util.AbstractList<String> {
    // superclass: generic_type -> scoped_type_identifier
    @Override
    public String get(int index) {
        return null;
    }

    @Override
    public int size() {
        return 0;
    }
}

// --- type_list (implements) variants ----------------------------------------

interface PlainIface {}

interface GenericIface<T> {}

class ImplementsPlain implements PlainIface {} // type_list: type_identifier

class ImplementsGeneric implements Comparable<ImplementsGeneric> {
    // type_list: generic_type -> type_identifier
    @Override
    public int compareTo(ImplementsGeneric o) {
        return 0;
    }
}

class ImplementsScoped implements java.io.Serializable {} // type_list: scoped_type_identifier

class ImplementsGenericScoped implements java.util.Comparator<String> {
    // type_list: generic_type -> scoped_type_identifier
    @Override
    public int compare(String a, String b) {
        return 0;
    }
}

// --- Type-defining declaration variants (java.tags.scm / java.types.scm) ---

class PlainClass {} // definition.class

interface PlainInterface {} // definition.interface

enum PlainEnum {
    // definition.enum; enum constant with constructor argument
    ONE(1),
    TWO(2);

    final int value;

    PlainEnum(int value) {
        this.value = value;
    }
}

record PlainRecord(int a, int b) {} // definition.class (record_declaration)

@interface PlainAnnotation {
    // definition.interface (annotation_type_declaration)
    String value();
}

// --- method_invocation variants (java.calls.scm) ----------------------------

class CallVariants {
    void plainCall() {
        identity(1); // no object field
    }

    void qualifiedCall() {
        Math.abs(-1); // object: identifier (Class.staticMethod())
    }

    void chainedCall() {
        String s = "x";
        s.trim().toUpperCase().length(); // object: method_invocation (chained)
    }

    int identity(int x) {
        return x;
    }
}

// --- Generic method / bounded type parameter --------------------------------

class GenericMethodHolder {
    <T extends Comparable<T>> T max(T a, T b) {
        return a.compareTo(b) >= 0 ? a : b;
    }
}

// --- Varargs -----------------------------------------------------------------

class VarargsHolder {
    void sum(int... nums) {
        int total = 0;
        for (int n : nums) {
            total += n;
        }
    }
}

// --- try-with-resources / switch (arrow form) / complexity variants --------

class ComplexityVariants {
    void tryWithResources() {
        try (var r = new java.io.StringReader("x")) {
            r.read();
        } catch (Exception e) {
            e.printStackTrace();
        } finally {
            System.out.println("done");
        }
    }

    String switchArrow(int n) {
        return switch (n) {
            case 1 -> "one";
            case 2 -> "two";
            default -> "other";
        };
    }

    void loops(int n) {
        for (int i = 0; i < n; i++) {
        }
        int i = 0;
        while (i < n) {
            i++;
        }
        do {
            i--;
        } while (i > 0);
        for (int x : new int[] {1, 2, 3}) {
        }
    }
}

// --- NEGATIVE cases: must not be captured as calls/definitions --------------

class NegativeHolder {
    int field;

    void negativeCases() {
        // A lambda is not a method_declaration; must never appear as
        // @definition.method / @definition.function in tags.
        Runnable lambdaBinding = () -> System.out.println("run");

        // A method reference is not an invocation; must never appear in
        // java.calls.scm's @call captures (it's a functional value, not a call).
        java.util.function.Supplier<String> methodRef = NegativeHolder::staticMethod;

        // Bare field access with no call parens must never appear as a call.
        int read = this.field;

        // Field write is not a "call".
        this.field = 5;
    }

    static String staticMethod() {
        return "x";
    }
}

File: crates/normalize-languages/tests/fixtures/c/variants.c
// Completeness fixture: one construct per grammar-legal variant of each
// field the c.{tags,calls,imports,types}.scm queries constrain, cross-
// referenced against arborium-c's node-types.json. Every construct here is
// *expected to be captured*; see query_fixtures.rs `c_*_completeness_*`
// tests for the matrix.
//
// This file also carries deliberate near-miss constructs (marked NEGATIVE)
// that must NOT be captured, to guard against over-broad patterns.

#include <stddef.h>

// --- struct_specifier / union_specifier name+body variants -----------------

struct PlainStruct {
    int x;
}; // struct_specifier: name = type_identifier, body present -> @definition.class

union PlainUnion {
    int i;
    float f;
}; // union_specifier: bare, name = type_identifier, body present ->
   // @definition.class. Previously entirely unmatched: the old query only
   // handled `declaration type: (union_specifier ...)`, which this bare
   // form never is.

typedef union {
    int i;
    float f;
} AnonUnion; // anonymous union_specifier (no name field) wrapped in a
             // typedef: the alias name "AnonUnion" is captured via the
             // type_definition pattern below; the union itself has no tag
             // name, correctly uncaptured (no name field to report).

typedef union TaggedUnion {
    int i;
    float f;
} TaggedUnionAlias; // named union_specifier nested inside a type_definition
                     // -> both "TaggedUnion" (union_specifier name+body) and
                     // "TaggedUnionAlias" (type_definition declarator) must
                     // be captured.

// --- function_declarator.declarator variants --------------------------------

int plain_function(int x) {
    return x; // declarator: identifier -> @definition.function
} // plain_function

int *pointer_returning_function(int x) {
    // Precedence: `*foo()` parses as "function returning pointer" —
    // pointer_declarator wraps function_declarator, so
    // function_declarator.declarator is still plain `identifier`.
    static int v;
    v = x;
    return &v;
}

int variadic_function(const char *fmt, ...) {
    // Variadic parameter does not change function_declarator.declarator —
    // still identifier. Included to document the construct is handled.
    (void)fmt;
    return 0;
}

int array_param_function(int arr[10]) {
    // `arr[10]` as a parameter is an array_declarator, but that's nested
    // inside parameter_list, not function_declarator.declarator itself
    // (still identifier "array_param_function").
    return arr[0];
}

// --- type_definition.declarator variants ------------------------------------

typedef struct {
    int *data;
} Handle; // plain typedef'd anonymous struct: declarator = type_identifier

typedef int (*FuncPtr)(int, int);
// Typedef'd function pointer: type_definition.declarator is
// function_declarator > parenthesized_declarator > pointer_declarator >
// type_identifier "FuncPtr" — three levels deeper than the plain case, and
// previously completely unmatched (verified via `normalize syntax query`
// against a probe file; this repo has no native C/C++ sources to measure
// density in, but callback-typedef idioms like this are ubiquitous in any
// C codebase with a callback/comparator API, e.g. POSIX signal handlers,
// qsort comparators, GLib/GTK signal callbacks).

// --- preproc_def / preproc_function_def (macro definitions) ----------------

#define MAX_SIZE 100
// preproc_def: name = identifier "MAX_SIZE" -> @definition.macro. Object-like
// macros had zero tags coverage before this fix.

#define SQUARE(x) ((x) * (x))
// preproc_function_def: name = identifier "SQUARE" -> @definition.macro.
// Function-like macros had zero tags coverage before this fix.

// --- enum_specifier ----------------------------------------------------------

enum Color { RED, GREEN, BLUE }; // name = type_identifier -> @definition.type

// --- nested/anonymous struct (documentation of existing correct behavior) --

struct Outer {
    struct {
        int x;
        int y;
    } point; // anonymous nested struct: no name field, correctly uncaptured
    union {
        int i;
        float f;
    }; // anonymous nested union member: same, correctly uncaptured
};

// --- multiple declarators in one declaration --------------------------------

int multi_a, multi_b, *multi_ptr;
// `declaration.declarator` is a multi-field; plain variable declarations
// aren't tags-worthy (consistent with every other language's convention of
// only tagging types/functions/macros, not bare variables), so none of
// these three produce a @name capture. Included to document the construct
// parses cleanly and doesn't trip up any pattern above.

// --- calls: field/pointer member call variants ------------------------------

void call_variants(void) {
    plain_function(1); // function: identifier -> @call
    FuncPtr fp = NULL;
    if (fp) {
        fp(1, 2); // call through a function-pointer variable: function:
                  // identifier "fp" -> @call (same shape as any plain call;
                  // no query change needed, included for documentation)
    }
}

struct WithFuncPtrField {
    FuncPtr fp;
};

int add_two(int a, int b) {
    return a + b;
}

void field_call_variants(struct WithFuncPtrField *p, struct WithFuncPtrField v) {
    p->fp(1, 2); // call_expression.function = field_expression, operator
                 // "->", field: field_identifier "fp" -> @call = "fp",
                 // @call.qualifier = "p"
    v.fp(3, 4);  // same shape, operator "."
    (void)add_two;
}

// --- NEGATIVE: constructs that must NOT match -------------------------------

int (*negative_function_pointer_variable)(int);
// `int (*fp)(int);` declares a variable of function-pointer type, not a
// function. function_declarator.declarator here is `parenthesized_declarator`
// (wrapping pointer_declarator -> identifier), never plain `identifier`, so
// this must NOT produce @definition.function. This is the correctly-excluded
// counterpart to the FuncPtr *typedef* case above (which legitimately should
// be captured, as a type alias, not a function).

union NegativeUsage;
void negative_union_usage_is_not_a_definition(void) {
    // A bodyless union-typed declaration references an existing tag; it must
    // never be reported as @definition.class (that would be true only when
    // the union_specifier also carries a body). This is exactly the false
    // positive the old `declaration type: (union_specifier ...)` pattern
    // produced (mislabeling a usage as @definition.class) and is now fixed
    // by requiring `body: (_)`.
}

struct NegativeHolder {
    int field;
};

int negative_field_access(struct NegativeHolder *holder) {
    // A bare field read through `->` with no call parens must never be
    // reported as a @call (regression guard against an over-eager
    // field_expression pattern matching plain member access).
    return holder->field;
}
